### PR TITLE
Only use HTTPS

### LINKED ISSUE #151: Start using https/SSL/TLS
Would be nice to have https/SSL/TLS support when communication with the Last.fm API.

With some mentoring insight into this codebase I'd be willing to contribute it.

### PR DESCRIPTION
**Pull request by simmel**
Closes #151

### CODE REVIEW

**Comment by hugovk** [comment]
There's a problem with Python 3.3: https://travis-ci.org/pylast/pylast/jobs/165224071#L216
and pypy: https://travis-ci.org/pylast/pylast/jobs/165224073#L216
and pypy3: https://travis-ci.org/pylast/pylast/jobs/165224074#L218

**Comment by simmel** [comment]
On Wed, 2016-10-05 at 06:51:23 -0700, Hugo wrote:
> There's a problem with Python 3.3: https://travis-ci.org/pylast/pylast/jobs/165224071#L216
> and pypy: https://travis-ci.org/pylast/pylast/jobs/165224073#L216
> and pypy3: https://travis-ci.org/pylast/pylast/jobs/165224074#L218

Looks like create_default_context() doesn't exist in Python <3.4 and we
have to do:
* Specify CA path
* Check that the domain we want to connect is listed in the certificate
that the server we connect to is present.

3.3 is still supported another year https://en.wikipedia.org/wiki/CPython#Version_history .
Looks like Ubuntu precise http://packages.ubuntu.com/search?keywords=python3 (which expires 2017-04) and CentOS via SCL https://www.softwarecollections.org/en/scls/rhscl/python33/ still uses 3.3. Although 3.4 can be installed via SCL https://www.softwarecollections.org/en/scls/rhscl/rh-python34/ .

What should we do about the 3.3 (and pypy3, sadly) users? Logging a warn
and be done with it? This feels like a bad idea but I'm not that excited
of writing our own wrapper for it, while searching I found pymongo's SSL
wrapper which is almost 400 LoC and importing one of two different CA
stores.

I have some code on my machine which I'm sure can work. I'm getting it
ready to commit within the day or so.


A side note: It looks like the supported URL for libre.fm now is
libre.fm/2.0 and not alpha.libre.fm? alpha.libre.fm will never work with
SSL (with CN/host validation on).

**Comment by hugovk** [comment]
Let's use HTTPS where available, and HTTP otherwise. That's an improvement over the current situation. 

http://www.randalolson.com/2016/09/03/python-2-7-still-reigns-supreme-in-pip-installs/

I'll update the Libre.fm URL.

**Comment by simmel** [comment]
On Fri, 2016-10-07 at 07:55:29 -0700, Hugo wrote:
> Let's use HTTPS where available, and HTTP otherwise. That's an improvement over the current situation. 

Where we can use it easily and securely by default?

Or where we can use it period? Because we can use it everywhere, it just
takes more code and dependencies.

> I'll update the Libre.fm URL.

Nice!

**Comment by simmel** [comment]
Thoughts Hugo?

I have both ideas coded I just need to know which one to push :)

**Comment by hugovk** [comment]
@simmel Everywhere sounds good!

**Comment by simmel** [comment]
@hugovk sadly I gave you false hope, using SSL securely everywhere isn't possible.

\>3.4 and >2.7.9 has sane defaults and just works(TM).

On 3.3 we have to use `certifi` and configure it explicitly to be secure. This sadly make all versions depend on `certifi` since I haven't found [a way in setuptools](http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-dependencies) or [requirements files](https://packaging.python.org/requirements/) to only depend on something when running under a specific Python version.

I think I'm done with this, can you review it?

**Comment by simmel** [comment]
Sweet! That SO link was exactly what I didn't find before! I settled for `extras_require` because it makes the code easier to remove and makes it stand out.

I hope I've fixed everything. This time, I ran pep8 two times before I commited ; )

File: pylast/__init__.py [hunk 1]
```diff
@@ -131,6 +149,34 @@ def _deprecation_warning(message):
...
+    ECDHE+SHA+AES+aRSA@STRENGTH
+    DHE+SHA+AES+aRSA@STRENGTH
+    """
+    cipher_string = ' '.join(cipher_string.split())
+    SSL_CONTEXT.set_ciphers(cipher_string)
+    SSL_CONTEXT.load_verify_locations(certifi.where())
+
+# Python >3.4 and >2.7.9 has sane defaults
+elif sys.version_info > (3, 4) or (
+        sys.version_info < (3, 0) and sys.version_info > (2, 7, 9)):
```
Comment (hugovk) [inline pylast/__init__.py]: Can this use `_can_use_ssl_securely()`?

The logic for `_can_use_ssl_securely()` is slightly different: 
```python
v > (3, 3) or (v < (3, 0) and v > (2, 7, 9))
```

Compared to here:
```python
v > (3, 4) or (v < (3, 0) and v > (2, 7, 9))
```

Are they intentionally different? They have similar comments:
> `>3.4 and >2.7.9 has sane defaults so use SSL there`

And:
> `Python >3.4 and >2.7.9 has sane defaults`

---

If we cannot reuse `_can_use_ssl_securely()`, let's use interval comparison:
```python
v > (3, 4) or ((2, 7, 9) < v < (3, 0))
```

File: pylast/__init__.py [hunk 2]
```diff
@@ -131,6 +149,34 @@ def _deprecation_warning(message):
 
 XML_ILLEGAL = re.compile(RE_XML_ILLEGAL)
 
+# Python <=3.3 doesn't support create_default_context()
+# <2.7.9 and <3.2 never did any SSL verification
+# FIXME This can be removed after 2017-09 when 3.3 is no longer supported and
+# pypy3 uses 3.4 or later, see
+# https://en.wikipedia.org/wiki/CPython#Version_history
+if sys.version_info < (3, 3, float("inf")) and sys.version_info > (3, 2):
```
Comment (hugovk) [inline pylast/__init__.py]: So this matches 3.2.x. Let's use:
```python
if sys.version_info[0] == 3 and sys.version_info[1] == 2:
```

File: pylast/__init__.py [hunk 3]
```diff
@@ -42,16 +42,34 @@
...
     warnings.warn(message, DeprecationWarning)
 
+
+def _can_use_ssl_securely():
+    # Python 3.3 doesn't support create_default_context() but can be made to
+    # work sanely.
+    # <2.7.9 and <3.2 never did any SSL verification so don't do SSL there.
+    # >3.4 and >2.7.9 has sane defaults so use SSL there.
+    v = sys.version_info
+    return v > (3, 3) or (v < (3, 0) and v > (2, 7, 9))
```
Comment (hugovk) [inline pylast/__init__.py]: We can use interval comparison:
```python
return v > (3, 3) or ((2, 7, 9) < v < (3, 0))
```

**Comment by simmel** [comment]
On Wed, 2016-10-19 at 04:48:29 -0700, Hugo wrote:
> @simmel Everywhere sounds good!

Ugh, sorry for spreading false hope.

<2.7.9 and <3.2 we can't do good how much we try. Should they use SSL
and hope for the best or just use HTTP there? Hope for the best is what
the code does now.

3.3 we can make sane via the `certifi` dependency and some
configuration, see the updated code. I'm looking into how to make us
ONLY depend on `certifi` when on Python 3.3.

>3.4 and >2.7.9 all is good.

File: pylast/__init__.py (hunk 3 above)
Comment (simmel) [inline pylast/__init__.py]: Looks like clever code but I'll allow it. Interval comparison looks neat (but clever)!

File: pylast/__init__.py (hunk 2 above)
Comment (simmel) [inline pylast/__init__.py]: 3.3 actually, but yes let's.

File: pylast/__init__.py (hunk 1 above)
Comment (simmel) [inline pylast/__init__.py]: No we can't since `_can_use_ssl_securely()` also includes 3.3.

Switched to clever code.

**Comment by simmel** [comment]
I'm clueless. TravisCI uses precise, an older Ubuntu version so I thought it might be a problem with too old OpenSSL but that doesn't seem to be the case: https://travis-ci.org/pylast/pylast/jobs/170998686#L198

Might be that pypy3 doesn't seem to have OpenSSL statically build in either but I'll try to list the ciphers in pypy3 some how.

**Comment by simmel** [comment]
Ah yes, ancient OpenSSL https://travis-ci.org/pylast/pylast/builds/171464271#L210

**Comment by simmel** [comment]
Should work now.

**Comment by simmel** [comment]
@hugovk they passed. On pypy3 the tests timed out though, any idea why? Looks like there were ~20ish tests left.

Looks like it always takes that long in pull requests https://travis-ci.org/pylast/pylast/builds/167208560
Your PR worked though https://travis-ci.org/pylast/pylast/jobs/165842936 I'm guessing it's because you have access to the secrets?

**Comment by hugovk** [comment]
Not sure what caused that pypy3 timeout.

Yes, my PR's have access to the secrets. One way is for you to add your encrypted credentials like another contributor has (see https://github.com/pylast/pylast/commit/f1e14f591c083b8bb1834aa9f8c5ce66e4d5a458), or if we do something from https://github.com/pylast/pylast/issues/171. 

Now Python 3.3 complains:

`NetworkError: [SSL: TLSV1_ALERT_DECODE_ERROR] tlsv1 alert decode error`

https://travis-ci.org/pylast/pylast/jobs/171465889#L368

**Comment by simmel** [comment]
On Sat, 29 Oct 2016, at 14:39, Hugo wrote:
> Not sure what caused that pypy3 timeout.
> Yes, my PR's have access to the secrets. One way is for you to add
> your encrypted credentials like another contributor has (see f1e14f5),
> or if we do something from #171[1].
> Just make those variables public

Sounds reasonable. In my "now playing"-script for irssi I've just made
them public since it's impossible not to. For me to require every user
to get an API-key is unreasonable.
> Now Python 3.3 complains:
> NetworkError: [SSL: TLSV1_ALERT_DECODE_ERROR] tlsv1 alert decode error
What the...? I'll look at it this evening. I didn't find that in the
pypy3 logs when I looked o_O


Links:

  1. https://github.com/pylast/pylast/issues/171

**Comment by hugovk** [comment]
Thanks. By the way, that error is from Python 3.3, not PyPy3.

**Comment by simmel** [comment]
@hugovk looks like this is why Mozillas SSL guide uses the expanded and not the compact version of the cipher strings.

**Comment by hugovk** [comment]
@simmel Thanks for all the work, merged!

**Comment by hugovk** [comment]
Released in pylast 1.7.0.

**Pull request merged**